src/v/cluster/tm_stm.h

```diff
@@ -400,6 +386,60 @@ class tm_stm final : public raft::persisted_stm<> {
     replicate_quorum_ack(model::term_id term, model::record_batch&& batch);
 
     model::record_batch serialize_tx(tx_metadata tx);
+
+    void upsert_transaction(tx_metadata);
+
+    fragmented_vector<tx_metadata> get_transactions_list() const;
+
+private:
+    std::chrono::milliseconds _sync_timeout;
+    config::binding<std::chrono::milliseconds> _transactional_id_expiration;
+    chunked_hash_map<model::producer_identity, kafka::transactional_id>
+      _pid_tx_id;
+    chunked_hash_map<kafka::transactional_id, ss::lw_shared_ptr<mutex>>
```

Comment:
I think locks could use some state consolidation here, this lifecycle is tied to _transaction entry, so can be merged into a single struct?

Reply:
i think i can do that in another PR, as now the transaction add is secured with a lock, this isn't going to be true when we would merge the two.